lib/AST/TypeJoinMeet.cpp

```diff
@@ -17,100 +17,197 @@
+  // There is no common superclass; we're done.
+  return nullptr;
+}
+
+Type TypeJoin::visitClassType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitBoundGenericClassType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitArchetypeType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitDynamicSelfType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitMetatypeType(Type second) {
+  assert(!First->mayHaveSuperclass() && !second->mayHaveSuperclass());
+
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
+    return nullptr;
+
+  auto firstInstance = First->castTo<AnyMetatypeType>()->getInstanceType();
+  auto secondInstance = second->castTo<AnyMetatypeType>()->getInstanceType();
+
+  auto joinInstance = join(firstInstance, secondInstance);
+
+  // FIXME: Return Any
+  if (!joinInstance)
+    return nullptr;
+
+  return MetatypeType::get(joinInstance);
+}
+
+Type TypeJoin::visitExistentialMetatypeType(Type second) {
+  assert(!First->mayHaveSuperclass() && !second->mayHaveSuperclass());
+
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
+    return nullptr;
+
+  auto firstInstance = First->castTo<AnyMetatypeType>()->getInstanceType();
+  auto secondInstance = second->castTo<AnyMetatypeType>()->getInstanceType();
+
+  auto joinInstance = join(firstInstance, secondInstance);
+
+  // FIXME: Return Any
+  if (!joinInstance)
+    return nullptr;
+
+  return ExistentialMetatypeType::get(joinInstance);
+}
+
+Type TypeJoin::visitBoundGenericEnumType(Type second) {
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
+    return nullptr;
+
   OptionalTypeKind otk1, otk2;
-  Type objectType1 = type1->getAnyOptionalObjectType(otk1);
-  Type objectType2 = type2->getAnyOptionalObjectType(otk2);
+  Type objectType1 = First->getAnyOptionalObjectType(otk1);
+  Type objectType2 = second->getAnyOptionalObjectType(otk2);
   if (otk1 == OTK_Optional || otk2 == OTK_Optional) {
     // Compute the join of the unwrapped type. If there is none, we're done.
-    Type unwrappedJoin = join(objectType1 ? objectType1 : type1,
-                              objectType2 ? objectType2 : type2);
+    Type unwrappedJoin = join(objectType1 ? objectType1 : First,
+                              objectType2 ? objectType2 : second);
+    // FIXME: More general joins of enums need to be handled.
     if (!unwrappedJoin) return nullptr;
 
     return OptionalType::get(unwrappedJoin);
   }
 
-  // The join can only be an existential.
+  // FIXME: More general joins of enums need to be handled, and
+  //        then Any should be returned when there is no better
+  //        choice.
   return nullptr;
 }
 
+Type TypeJoin::visitOptionalType(Type second) {
+  auto canFirst = First->getCanonicalType();
+  auto canSecond = second->getCanonicalType();
+
+  return TypeJoin::join(canFirst, canSecond);
```

Comment:
You want to look at the underlying object type, I presume.

Reply:
Uh...yeah. No tests fail so I'll look into whether I can come up with something.

Reply:
Oh wait no, this is just calling back into join with the desugared types to dispatch to the `BoundGenericEnum` case.